Add OpOverload.redispatch; use it in new custom ops API

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124299
* #124200
* #124199
* #124134
* #124180
* #124092
* __->__ #124089
* #124071
* #124066
* #124065
* #124064
* #123937

A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

zou3519 committed:
Add OpOverload.redispatch; use it in new custom ops API

A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

albanD commented on torch/_library/autograd.py:
```diff
@@ -48,7 +55,13 @@ def backward(ctx, *grads):
 
-    def autograd_impl(*args):
+    def autograd_impl(keyset, *args):
+        # We set a nonlocal to ferry keyset from here to the forward.
```
Why not pass it as an argument?

zou3519 replied:
this changes `ctx.needs_input_grad` and I don't want to fix that right now haha

albanD commented on torch/_library/custom_ops.py:
```diff
@@ -392,12 +392,12 @@ def fake_impl(*args, **kwargs):
 
         autograd_impl = _library.autograd.make_autograd_impl(self._opoverload, self)
-        lib.impl(self._name, autograd_impl, "Autograd")
+        lib.impl(self._name, autograd_impl, "Autograd", dispatcher_convention=True)
```
Other naming proposals:
- with_keyset
- redispatch_convention
- pass_keyset

Also should we do what we do in c++ where we inspect the first argument name and if it is name keyset, we use this convention automatically? :p

zou3519 replied:
> Also should we do what we do in c++ where we inspect the first argument name and if it is name keyset, we use this convention automatically? :p

I'm happy to do this, was worried about if it was too BC-breaking.

zou3519 replied:
> Also should we do what we do in c++ where we inspect the first argument name and if it is name keyset, we use this convention automatically? :p

I changed my mind, this is risky because inspect.signature doesn't handle all cases so we'd likely break BC.

albanD reviewed (commented):
Pretty cool!

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

albanD reviewed (approved):
Nice!

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add OpOverload.redispatch; use it in new custom ops API"


A kernel has "dispatcher convention" if there is an additional keyset
arg at the beginning of the argument list. This PR:
- adds a way to register kernels with dispatcher_convention using
  Library.impl (pass dispatcher_convention = True)
- adds OpOverload.redispatch

We use both of the above in the new custom ops API: we register the
autograd kernel in dispatcher convention so that we can actually call
redispatch like how pytorch built-in ops do it.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
